### DISCUSSION TITLE
Positive resist  electron beam lithography

### DISCUSSION THREAD

**Discussion by DorianHGrey** [gdsfactory/gdsfactory#753] [Q&A] [answered]
Hello guys! Is there a way to generate gds file or mask for positive resist? Tried gf.outline on grating parts.  Did not get what I want. the first figure(rectangle grating ) was pattern in negative resist. To generate such structure in the SOI with positive resist(EBL), It needed some modification and turn into something like second one.With bbox and boolean subtraction, it could be done. But the exposure time would be too long and cost a lot. If there is better way, please share with me. Thanks a lot.

![Screenshot 2022-10-05 at 4 10 21 PM](https://user-images.githubusercontent.com/114927601/194012493-54c830fd-10d6-47f8-b8f6-5047e4542774.png)
![Screenshot 2022-10-05 at 4 10 37 PM](https://user-images.githubusercontent.com/114927601/194012540-a6e30c8d-a1e1-4460-8176-6c3d4e8216ed.png)

**Comment by joamatab** [comment]
Hi, can you add the code that you used and explain a bit more of what you need?

To help you we need to be able to understand and reproduce your experience

**Reply by DorianHGrey** [reply]
Thank you for your help. I uploaded some figure and detailed the question to make it more clear.If you know the solution, please share with me. Thanks a lot.

**Comment by joamatab** [comment, marked as answer]
Hi Dorian,

the easiest way is to define the core and the cladding of your waveguides and do a final mask preparation step

https://github.com/atait/lymask

**Comment by jan-david-fischbach** [comment]
With the new `gdstk` backend we have actually been able to do the booleans in gdsfactory itself (sufficiently fast), which might make for a more "monolithic" design flow.

``` python
@gf.cell
def invert_positive_layers(
    old: gf.Component, dimensions: Tuple[float, float]
) -> gf.Component:
    """
    Invert all layers which have a datatype equal to 4 (used to indicate a positive resist layer) 

    Parameters:
     * old (gf.Component): The component that should have the concerned layers inverted
     * dimensions (Tuple[float, float]): outer dimensions of the design `old` dictating to which extent the layer should be inverted

    Returns:
     * c (gf.Component): New cell with the applied inversion. Due to the boolean operation the child-cells will be flattened
    """
    c = gf.Component()

    positive_layers = []
    negative_and_virtual = []
    for layer in LAYER:
        if layer[1][1] == 4:  # layer is positive-resist as seen from the datatype
            positive_layers.append(
                layer[1]
            )  # this is important, as extract only takes the layer tuple as input
        else:
            negative_and_virtual.append(layer[1])

    outer = gf.components.rectangle(dimensions, centered=True)
    for layer in positive_layers:
        extract_n = old.extract(layers=(layer,))
        c << gf.geometry.boolean(outer, extract_n, "A-B", layer=layer)

    extract_p = old.extract(layers=negative_and_virtual)

    c << extract_p
    return c

```

**Reply by 233thzhang** [reply]
> 有了新的`gdstk`后端，我们实际上已经能够在gdsfactory本身中执行布尔值（速度足够快），这可能会形成一个更“单一”的设计流程。
> 
> ```python
> @gf.cell
> def invert_positive_layers(
>     old: gf.Component, dimensions: Tuple[float, float]
> ) -> gf.Component:
>     """
>     Invert all layers which have a datatype equal to 4 (used to indicate a positive resist layer) 
> 
>     Parameters:
>      * old (gf.Component): The component that should have the concerned layers inverted
>      * dimensions (Tuple[float, float]): outer dimensions of the design `old` dictating to which extent the layer should be inverted
> 
>     Returns:
>      * c (gf.Component): New cell with the applied inversion. Due to the boolean operation the child-cells will be flattened
>     """
>     c = gf.Component()
> 
>     positive_layers = []
>     negative_and_virtual = []
>     for layer in LAYER:
>         if layer[1][1] == 4:  # layer is positive-resist as seen from the datatype
>             positive_layers.append(
>                 layer[1]
>             )  # this is important, as extract only takes the layer tuple as input
>         else:
>             negative_and_virtual.append(layer[1])
> 
>     outer = gf.components.rectangle(dimensions, centered=True)
>     for layer in positive_layers:
>         extract_n = old.extract(layers=(layer,))
>         c << gf.geometry.boolean(outer, extract_n, "A-B", layer=layer)
> 
>     extract_p = old.extract(layers=negative_and_virtual)
> 
>     c << extract_p
>     return c
> ```

Hello Jan-David-Black, 
I got confused when running this code, can you provide a more complete example of the code, for example with the core ,cladding and final export  layers, please.

**Reply by joamatab** [reply]
Hi, We have made a new tutorial for this

https://gdsfactory.github.io/gdsfactory/notebooks/dataprep.html

which complements this one
https://gdsfactory.github.io/gdsfactory/notebooks/04_components_geometry.html

and you can use the `add_tenches` Pcell to add trenches to any component
https://gdsfactory.github.io/gdsfactory/components.html#add-trenches


@elamdf
@mdecea 
@simbilod 
@HelgeGehring 
@Djordje117 
@bhnzhang

**Reply by 233thzhang** [reply]
Thank you sir,

I have tried  the tutorial  that you listed. But there still have some wrong that could not run smoothly.

When I run the code in  tutorial , like this one 
`from gdsfactory.generic_tech.layer_map import LAYER as l
import gdsfactory as gf

gf.config.rich_output()
PDK = gf.generic_tech.get_generic_pdk()
PDK.activate()

from functools import partial
import gdsfactory as gf
from gdsfactory.geometry.maskprep import get_polygons_over_under, over_under

c = gf.components.coupler_ring(
    cladding_layers=((2, 0)),
    cladding_offsets=(0.2,),
)
c`

and the output shows that "No module named 'gdsfactory.geometry.maskprep'"

---------------------------------------------------------------------------
ModuleNotFoundError                       Traceback (most recent call last)
Cell In[1], line 10
      8 from functools import partial
      9 import gdsfactory as gf
---> 10 from gdsfactory.geometry.maskprep import get_polygons_over_under, over_under
     12 c = gf.components.coupler_ring(
     13     cladding_layers=((2, 0)),
     14     cladding_offsets=(0.2,),
     15 )
     16 c

ModuleNotFoundError: No module named 'gdsfactory.geometry.maskprep'

After that, I checked the func  "gdsfactory.components.add_trenches" ,and the output are similar, No module named  "gdsfactory.components.add_trenches" 

To find out where is the key point, i also check the version of gdsfactory version in Anaconda Promot,  it shows the version is 6.64.2, and i use the command "pip install gdsfactory --upgrade". The version of gdsfactory did not change, stiil is 6.64.2. And the newest version of gdsfactory is 6.73.1, I am not sure does it cause the issue?

By the way, the version of python of anaconda environment is 3.9.16. 

I hope these information could help us to find where is the key point of this issue.

**Reply by joamatab** [reply]
yes, that feature was released yesterday so you need the latest gdsfactory version

make sure you use the anaconda prompt with the activated environment

see instructions here
https://gdsfactory.github.io/gdsfactory/

you can also delete your python and install a new python with latest gdsfactory


This code will tell you which gdsfactory you are using
```
import gdsfactory as gf

gf.config.print_version()
```
